### PR TITLE
Improved CMND implementation for YIN / pYIN

### PR DESCRIPTION
**Pull request by dsuedholt**
**TL;DR**: We can noticeably improve the accuracy of the YIN implementation by getting rid of the `win_length` parameter entirely and simply calculating the autocorrelation function (ACF) over the entire frame. Mathematical reasoning and some benchmarks provided below.

Apologies in advance for the wall of text. The devil is very much in the details with this one and I wanted to document it thoroughly.

## ACF Calculation

The main point of this pull request is to replace the calculation of the ACF: https://github.com/librosa/librosa/blob/d5aa7e1ae443c917fa53bef31055ed2613d63133/librosa/core/pitch.py#L393-L395 with a call to librosa's existing implementation: 

```python
acf_frames = audio.autocorrelate(y_frames, max_size=max_period+1, axis=-2)
``` 

Looking at this [discussion](https://github.com/librosa/librosa/pull/1063#discussion_r435547835), I believe the intention behind the current implementation is to calculate a type I ACF with a constant window size. Given window length $N$, this is defined as

$$r(k) = \sum_{n=0}^{N-1}x[n]x[n + k]$$

By contrast, the type II ACF uses an integration window that shrinks as the lag increases:

$$r'(k) = \sum_{n=0}^{N-1-k}x[n]x[n + k]$$

I'll go into some detail as to why I believe that 

1. the current implementation does not actually correctly calculate the type I ACF
2. the librosa implementation should use the type II ACF anyway (as implemented in `librosa.core.autocorrelate`)

It's true that the authors of the original YIN paper recommend to use the type I ACF, because the type II ACF implicitly tapers to 0 as $k \rightarrow N$. However, in the formulation of the type I ACF, the $n + k$ index will exceed the window size. This is not an issue discussed in the paper, since the authors consider the signal in its entirety, rather than on a frame-by-frame basis, so they can essentially "look into the future". 

As I understand it, the current implementation addresses this by making a distinction between `frame_length` $N$ and `win_length` $W < N$, where `win_length` essentially places an upper bound on the calculated lags. Mathematically,

$$r(k) = \sum_{n=0}^{W-1}x[n]x[n+k] \text{ for } 0\leq k < N - W$$

However, due to the `win_length:0:-1` indexing, the current code instead calculates:

$$r(k) = \sum_{n=0}^{W-1}x[n]x[n+k+1] \text{ for } 0\leq k < N - W$$

This off-by-one error was introduced [here](https://github.com/librosa/librosa/issues/1425). I think the reason that this change actually improved accuracy at the time is that it might have cancelled out another off-by-one error in the following calculation of the cumulative mean normalized difference (CMND) function, which this PR also fixes (see next section).

But this brings me to the second point of the argument: rather than fixing this calculation, I propose to drop the `win_length` parameter entirely and simply use `librosa.core.audio.autocorrelate` to calculate the type II ACF over the entire frame. This has several benefits:

- it simplifies the function call interface by removing a parameter
- using the type II ACF is standard in frame-by-frame implementations of YIN, in e.g. [aubio](https://github.com/aubio/aubio/blob/master/src/pitch/pitchyin.c) and also in the [original pYIN implementation](https://code.soundsoftware.ac.uk/projects/pyin/repository/entry/YinUtil.cpp#L57), so it aligns librosa with expected behavior.

As an example, the different ACFs are illustrated below for a 2048-sample frame from track 2 of the vocadito dataset. The highest peak of the type II ACF (excluding 0 lag) is at the correct period, while the type I ACF's peak at the correct period is slightly lower than the peak at double the period.

ACF comparison

## CMND Calculation

The next step in the YIN algorithm is to calculate the difference function. As mentioned before, the current implementation makes another off-by-one error here: https://github.com/librosa/librosa/blob/d5aa7e1ae443c917fa53bef31055ed2613d63133/librosa/core/pitch.py#L398-L406

Translating line 401 into mathematical notation, we see that the `energy_frame` $E$ is calculated as

$$\begin{align}
E(k) &= \sum_{n=0}^{W + k}x[n]^2 - \sum_{n=0}^{k}x[n]^2\\
&= \sum_{n=k+1}^{W + k}x[n]^2\\
&= \sum_{n=0}^{W - 1}x[n + k + 1]^2
\end{align}$$

Meaning that we end up with the same off-by-one error as seen before in the ACF correlation, which seem to cancel each other out to a degree.

The calculation needs to be adapted in any case to account for switching to the type II ACF $r'(k)$. We'll define 

$$\begin{align}E(k) &= \sum_{n=k}^{N-1}x[n]^2\\
&= \sum_{n=0}^{N-1}x[n]^2 - \sum_{n=0}^{k-1}x[n]^2\\
&= r'(0) - \sum_{n=0}^{k-1}x[n]^2
\end{align}$$

Following the logic of equation (7) from the YIN paper, the difference function can be calculated as

$$\begin{align}
d(k) &= r'(0) + E(k) - 2r'(k)\\
&= r'(0) + \left(r'(0) - \sum_{n=0}^{k-1}x[n]^2\right) - 2r'(k)\\
&= 2\left(r'(0) - r'(k)\right) - \sum_{n=0}^{k-1}x[n]^2
\end{align}$$

And going back from maths to code:

```python
yin_frames = np.cumsum(np.square(y_frames), axis=-2)

# Difference function: d(k) = 2 * (ACF(0) - ACF(k)) - sum_{m=0}^{k-1} y(m)^2
k = slice(1, max_period + 1)
yin_frames[..., 0, :] = 0
yin_frames[..., k, :] = (
    2 * (acf_frames[..., 0, :] - acf_frames[..., k, :]) - yin_frames[..., :k.stop-1, :]
)
``` 

The remaining steps (cumulative mean normalization, peak picking etc) are correct in the current implementation as far as I can tell.

## Benchmarks

I've run some comparisons using `mir_eval` that hopefully show that all this mathematical pedantry was worth it. The full numbers are contained [this Google sheet](https://docs.google.com/spreadsheets/d/15_eYfSo3drQSCM-cWmeYEokGsPsMtqqpg4E5PhnP5bs/edit?gid=0#gid=0), but I'll summarize the highlights here.

Using `fmin=80`, `fmax=800`, `sr=44100` and leaving all other values at their default, the new implementation improves the raw pitch accuracy of YIN (using a tolerance of 50 cents) on the [vocadito](https://zenodo.org/records/5578807) dataset significantly:

- from 95% to 98% using `frame_length=1024`
- from 92% to 98% using `frame_length=2048`

pYIN results also show clear improvements both for raw pitch accuracy and overall accuracy (taking into account V/UV decisions). 

The drawback is an increase in computational cost. The FFT is $O(N\log(N))$, and we've replaced 3 FFT operations with length $N/2$ with 2 operations of length $N$. That means at $N=2048$, we'd expect to take roughly 1.5x longer, which indeed matches the measurements for YIN. However, due to the more costly Viterbi decoding, this does not seem to have the same effect on pYIN running times.

## Additional changes

- I removed this line in both YIN and pYIN which marks $d'(0)$ as a trough if $d'(0) < d'(1)$: https://github.com/librosa/librosa/blob/d5aa7e1ae443c917fa53bef31055ed2613d63133/librosa/core/pitch.py#L612 This check is unnecessary since $d'(0) = 1$ by definition (equation 8 in the YIN paper), and a lag of 0 should never be selected as the period anyway.
- `__check_yin_parameters`
  - Since `win_length` is unused now, I've added a deprecation warning if it is called with any value other than `None`
  - It previously included this check: https://github.com/librosa/librosa/blob/d5aa7e1ae443c917fa53bef31055ed2613d63133/librosa/core/pitch.py#L967 I believe this is an error and should use `fmin`? As I understand it, the intention is an upper bound on the largest possible period, which corresponds to the lowest frequency, not the highest. I've changed this check to throw an error if the largest period `sr / fmin` exceeds `frame_length`.
  - Additionally there's a "softer" check (warning instead of error) for `sr / fmin >= frame_length // 2`, since ideally the frame should fit two periods of the lowest frequency for accurate f0 detection.

### CODE REVIEW

**Comment by bmcfee** [comment]
Thanks for this!  Dropping a note now to acknowledge that it's on my radar, but I won't have time to dig into it for a week or two.

File: librosa/core/pitch.py [hunk 1]
```diff
@@ -959,15 +946,28 @@ def __check_yin_params(
     if fmin <= 0:
         raise ParameterError(f"fmin={fmin:.3f} must be strictly positive")
 
-    if win_length >= frame_length:
+    if sr / fmin >= frame_length - 1:
```
Comment (lostanlen) [inline librosa/core/pitch.py:949]: does this PR introduce `ParameterError`'s for cases which were did not error (at least for some values of `win_length`) up until today?

Comment (dsuedholt) [reply librosa/core/pitch.py:949]: In the current state of this PR, `win_length` is actually completely ignored. A deprecation warning (not an error) will be logged if `yin` or `pyin` is called with `win_length` set to any value other than `None`.

I do think the `win_length` parameter should be discarded entirely going forward, and I think this way it's not breaking anyone's existing code.

However, I think it might introduce `ParameterError`s for some values of `fmin` and `fmax`. I believe the current librosa code erroneously checks `fmax` to determine whether the largest period fits into the frame, which this PR changes to checking `fmin` (lowest frequency ≘ largest period).

**Comment by bmcfee** [comment]
:wave: thanks again @dsuedholt for the thorough investigation and writeup of your changes here.  Also thanks for your patience while I cleared my stack enough to put some thought into this PR!

Everything you've described makes total sense, and indeed clarifies a few points that popped up in the development of this implementation.

I've enabled the CI to run, and it looks like the changes here do introduce several regressions.  Some of these are down to changes in estimation - I suspect the disabled trough check here  - and others are due to broken handling of multi-channel.  I'll try to sleuth the latter part out to recommend a fix.



> * I believe this is an error and should use `fmin`? As I understand it, the intention is an upper bound on the largest possible period, which corresponds to the lowest frequency, not the highest. I've changed this check to throw an error if the largest period `sr / fmin` exceeds `frame_length`.

:eyes: Huh  We had this right in the issue thread where this was all derived: https://github.com/librosa/librosa/issues/1680#issuecomment-1487241064 ; this was then reverted in the PR at commit 6923fe70a96883a9f1c5b9378ce9dc62d4d1fa25 but I can't at all remember why this was a problem.

File: librosa/core/pitch.py (hunk 1 above)
Comment (bmcfee) [reply librosa/core/pitch.py:949]: If `win_length` is deprecated, please use the `util.deprecation.Deprecated` helper class to mark it as such.  This way we can distinguish between an explicit `None` setting (which would be an error going forward) and using the default (which would not be an error in the future).

File: librosa/core/pitch.py [hunk 2]
```diff
@@ -597,19 +584,18 @@ def yin(
...
-        y_frames, frame_length, win_length, min_period, max_period
+        y_frames, min_period, max_period
     )
 
     # Parabolic interpolation.
     parabolic_shifts = _parabolic_interpolation(yin_frames)
 
     # Find local minima.
     is_trough = util.localmin(yin_frames, axis=-2)
-    is_trough[..., 0, :] = yin_frames[..., 0, :] < yin_frames[..., 1, :]
```
Comment (bmcfee) [inline librosa/core/pitch.py:612]: I think this line is still needed; it's here to ensure that the localmin calculation is properly defined at the edge of the array.

File: librosa/core/pitch.py [hunk 3]
```diff
@@ -773,10 +762,6 @@ def pyin(
     if fmin is None or fmax is None:
         raise ParameterError('both "fmin" and "fmax" must be provided')
 
-    # Set the default window length if it is not already specified.
-    if win_length is None:
-        win_length = frame_length // 2
-
     __check_yin_params(
```
Comment (bmcfee) [inline librosa/core/pitch.py:765]:
```suggestion
   if not isinstance(win_length, Deprecated):
       warnings.warn(FutureWarning, "win_length parameter is deprecated...")
```

File: librosa/core/pitch.py [hunk 4]
```diff
@@ -885,7 +870,6 @@ def __pyin_helper(
         # 2. For each frame find the troughs.
         is_trough = util.localmin(yin_frame)
 
-        is_trough[0] = yin_frame[0] < yin_frame[1]
```
Comment (bmcfee) [inline librosa/core/pitch.py:888]: as above, i think this part may still be needed

File: librosa/core/pitch.py [hunk 5]
```diff
@@ -389,31 +384,25 @@ def _cumulative_mean_normalized_difference(
...
-    acf_frames[np.abs(acf_frames) < 1e-6] = 0
+    acf_frames = audio.autocorrelate(y_frames, max_size=max_period + 1, axis=-2)
 
     # Energy terms.
-    energy_frames = np.cumsum(y_frames**2, axis=-2)
-    energy_frames = (
-        energy_frames[..., win_length:, :] - energy_frames[..., :-win_length, :]
-    )
-    energy_frames[np.abs(energy_frames) < 1e-6] = 0
+    yin_frames = np.cumsum(np.square(y_frames), axis=-2)
```
Comment (bmcfee) [inline librosa/core/pitch.py:390]: not for this PR, but cumsum(square) composition seems like a good candidate for numba ufunc optimization.

File: librosa/core/pitch.py [hunk 6]
```diff
@@ -389,31 +384,25 @@ def _cumulative_mean_normalized_difference(
...
     yin_numerator = yin_frames[..., min_period : max_period + 1, :]
     # broadcast this shape to have leading ones
-    tau_range = util.expand_to(
-        np.arange(1, max_period + 1), ndim=yin_frames.ndim, axes=-2
-    )
+    k_range = util.expand_to(np.r_[k], ndim=yin_frames.ndim, axes=-2)
 
     cumulative_mean = (
-        np.cumsum(yin_frames[..., 1 : max_period + 1, :], axis=-2) / tau_range
+        np.cumsum(yin_frames[..., k, :], axis=-2) / k_range
```
Comment (bmcfee) [inline librosa/core/pitch.py:405]: somewhere in this passage, I think, is where the multichannel failures are creeping in.  @dsuedholt could you step this on a multi-channel example and make sure that everything retains the proper shapes compared to the current main branch?

File: librosa/core/pitch.py (hunk 2 above)
Comment (dsuedholt) [reply librosa/core/pitch.py:612]: I suppose it's a question of definition whether rigorously speaking, an edge point should be considered a local minimum or not. However, I don't think it makes a practical difference in this case. The algorithm will only pick lag candidates within the search bounds, and a lag of 0 can't be within those bounds since it would correspond to "infinite frequency". 

The `localmin` call returns a boolean array, where `is_trough[..., 0, :] == False` by definition according to the [`localmin` documentation](https://librosa.org/doc/main/generated/librosa.util.localmin.html) - I think this suits the algorithm well and doesn't require that additional check.

But I don't think the extra check impacts performance or accuracy, so I'd be happy to add it back in.

Comment (bmcfee) [reply librosa/core/pitch.py:612]: > and a lag of 0 can't be within those bounds since it would correspond to "infinite frequency".

That's true, but `yin_frames` is already trimmed to the min / max period before returning from CNMD, so "0" here actually means min_period.

Comment (dsuedholt) [reply librosa/core/pitch.py:612]: Ah my bad, you're completely right of course! I'll revert this change.

**Comment by dsuedholt** [comment]
Hi @bmcfee, thanks a lot for the helpful review! Multichannel processing should be fixed, tests are passing now.

I couldn't find the code that generated the ground truth data for the chirps that's being loaded from `tests/data/pitch-yin.npy` 🤔 I changed the tests to use the average instantaneous frequency for every frame as ground truth instead.

**Comment by bmcfee** [comment]
> I couldn't find the code that generated the ground truth data for the chirps that's being loaded from `tests/data/pitch-yin.npy`

I believe these outputs came from a prior yin implementation (from the vamp plugin, IIRC) so that we could regression-test for compatibility; see discussion in https://github.com/librosa/librosa/pull/1063 .  I'd prefer to keep that test as it was so that we don't lose track of compatibility.  If we need to relax the test somehow to account for differences introduced in this change, we can do that, but retain documentation (comments) about what changed and why.

I'm happy to have a free-standing unit test against the instantaneous frequency as a separate test though.

Thanks again for all this!

**Comment by dsuedholt** [comment]
Makes sense! I got the previous tests against the vamp plugin working now with a few adjustments. I reduced the framesize in the YIN call to 1024; before it was being called with the default `frame_length=2048`, but because of the previous `win_length` implementation, it was effectively only looking at 1024 samples.

For the `pyin` chirp test, I switched to the exact values used in the vamp plugin for fmin, fmax, and resolution, and it seems to pass the test for me.

I did notice one more thing though: the tests assumed that the first frame of the pyin output will always be unvoiced, because the decoding is initialized with a prior of 0 for all voiced bins and `1 / n_pitch_bins` for all unvoiced bins, instead of a uniform `1 / (2 * n_pitch_bins)` for all bins: https://github.com/librosa/librosa/blob/d5aa7e1ae443c917fa53bef31055ed2613d63133/librosa/core/pitch.py#L853-L854

I'm not sure why this decision was made, but it matches neither the original paper (defines uniform V/UV prior in equation 6) nor the Vamp plugin (initializes a uniform prior [here](https://code.soundsoftware.ac.uk/projects/pyin/repository/entry/MonoPitchHMM.cpp#L91)). 

I changed the pyin implementation to a uniform prior, which made the `test_pyin_multi_center` test pass for me when it previously didn't (I'm on a mac, I'm not sure if the mentioned OSX issue still persists).

**Comment by dsuedholt** [comment]
I added tests for the warnings (should fix reduced code coverage) and removed a `numpy>2.0` only feature from one of the new chirp tests (should fix the linter failing).

I'm not sure what to do about the doc workflow failing, I can't get the error to happen on my local machine.

File: tests/test_core.py [hunk 7]
```diff
@@ -1171,43 +1171,74 @@ def test_yin_tone(freq):
...
     y = librosa.tone(freq, duration=1.0)
     f0, _, _ = librosa.pyin(y, fmin=110, fmax=1000, center=False)
-    # Skip the first frame, since the voicing prior does not allow it
-    assert np.allclose(np.log2(f0[1:]), np.log2(freq), rtol=0, atol=1e-2)
+    assert np.allclose(np.log2(f0), np.log2(freq), rtol=0, atol=1e-2)
 
 
 def test_pyin_multi():
-    y = np.stack([librosa.tone(440, duration=1.0), librosa.tone(560, duration=1.0)])
+    y = np.stack([librosa.tone(440, duration=0.1), librosa.tone(560, duration=0.1)])
```
Comment (bmcfee) [inline tests/test_core.py:1241]: I'm curious why this duration was shortened so much in this test?

**Comment by bmcfee** [comment]
> I added tests for the warnings (should fix reduced code coverage) and removed a `numpy>2.0` only feature from one of the new chirp tests (should fix the linter failing).
> 

Thanks for this - it's all looking great!  I had one minor q about the revised tests, but I'm otherwise happy with where things are at.

> I'm not sure what to do about the doc workflow failing, I can't get the error to happen on my local machine.

The doc CI was failing due to a broken link.  This is almost surely unrelated to the PR, and it happens sporadically when sites we link to are temporarily unreachable.  As long as the tests that execute before link check are passing, I won't worry about it.

File: librosa/core/pitch.py (hunk 5 above)
Comment (bmcfee) [reply librosa/core/pitch.py:391]: Actually, we can cut memory consumption in half here with an in-place cumsum.  No further optimization needed.
```suggestion
    yin_frames = np.square(y_frames)
    np.cumsum(yin_frames, out=yin_frames, axis=-2)
```

**Comment by bmcfee** [comment]
> I'm not sure why this decision was made, but it matches neither the original paper (defines uniform V/UV prior in equation 6) nor the Vamp plugin (initializes a uniform prior [here](https://code.soundsoftware.ac.uk/projects/pyin/repository/entry/MonoPitchHMM.cpp#L91)).

That's also a good point.  It must have escaped our attention in the initial CR of #1063 because the test signals generally started on unvoiced frames.  I couldn't find any discussion of it in the comment thread anyway.

File: tests/test_core.py (hunk 7 above)
Comment (dsuedholt) [reply tests/test_core.py:1241]: Oops, sorry. That's a debugging remnant that shouldn't have been committed, I'll change that back.

File: librosa/core/pitch.py (hunk 5 above)
Comment (dsuedholt) [reply librosa/core/pitch.py:391]: Thanks! I'll add in this change.

**Review by bmcfee — APPROVED** [review]
LGTM

**Comment by bmcfee** [comment]
I think we're all set with this.  I gave it a spin locally and it's all behaving nicely.

We do take a bit of a performance hit here: speed was about the same, but memory usage is up by about 10%.  I expect this comes from some inefficiencies in either `autocorrelate` or `abs2` (or both), relative to how it was previously implemented.  I can live with this for now - we can always revisit internal optimizations in a later point release.

Unless there are objections, I think this is good to merge.  Thanks again @dsuedholt for the deep dive and scrutiny on this!
